app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabase.kt

```diff
@@ -463,6 +488,86 @@ class FirebaseDatabase : Database() {
         return Utils.failedFuture(Error("Not implemented"))
     }
 
+    override fun getTournamentUID(): String? {
+        return database.child(FirebaseKeys.TOURNAMENTS_ROOT).push().key
+    }
+
+    override fun addTournament(tournament: Tournament): CompletableFuture<Unit> {
+        val future = CompletableFuture<Unit>()
+        // add the tournament to all tournaments
+        tournamentsRoot().updateChildren(mapOf(tournament.id to tournament))
+            .addOnSuccessListener { future.complete(Unit) }
+            .addOnFailureListener { future.completeExceptionally(it) }
+        return future
+    }
+
+    override fun removeTournament(tournamentId: String): CompletableFuture<Unit> {
+        val future = CompletableFuture<Unit>()
+
+        database.child("${FirebaseKeys.TOURNAMENTS_ROOT}/$tournamentId/${FirebaseKeys.TOURNAMENT_PARTICIPANTS_IDS}")
+            .get()
+            .addOnSuccessListener { data ->
+                // get the list of all participants (list of ids)
+                val participantsIds = transformSnapshotKeysToStringList(data)
+                // prepare changes in database
+                // 1. remove the tournament from the list of tournaments of all participants
+                val changes: MutableMap<String, Any?> = participantsIds.associate {
+                    "${FirebaseKeys.USERS_ROOT}/$it/${FirebaseKeys.USER_TOURNAMENTS}/$tournamentId" to null
+                }.toMutableMap()
+                // 2. remove the tournament from the tournaments file
+                changes.put("${FirebaseKeys.TOURNAMENTS_ROOT}/$tournamentId", null)
+                // do the changes
+                database.updateChildren(changes).addOnSuccessListener { future.complete(Unit) }
+                    .addOnFailureListener { future.completeExceptionally(it) }
+            }
+            .addOnFailureListener { future.completeExceptionally(it) }
+
+        return future
+    }
+
+    override fun addUserToTournament(
+        userId: String,
+        tournamentId: String,
+    ): CompletableFuture<Unit> {
+        val future = CompletableFuture<Unit>()
+        // check that the userId and tournamentId exist
+        isUserInDatabase(userId).thenCombine(isTournamentInDatabase(tournamentId)) { userExists, tournamentExists ->
+            run {
```

Comment:
Why use `run` here?

Apart from that, this method is really well implemented with a good use of the `Future` API. Also nice to do the two writes in one go 👍.

Reply:
I forgot to remove it, good catch. 
And thank you, I've had a hard time figuring these futures out!